File: src/server/services/response.service.ts
import { appendAuditBlock, readDb, writeDb } from '@/server/db/database';
import { ContainmentAction, ContainmentType } from '@/shared/types';

export async function getAllContainmentActions(): Promise<ContainmentAction[]> {
  const db = readDb();
  return db.containmentActions || [];
}

export async function executeContainmentAction(data: {
  type: ContainmentType;
  targetId: string;
  targetName: string;
  executedBy?: string;
  reason?: string;
}): Promise<ContainmentAction> {
  const db = readDb();
  if (!db.containmentActions) db.containmentActions = [];

  const executedBy = data.executedBy || 'SOC Operator (CipherNest Desktop)';
  const details = data.reason || `Automated containment rule executed for target: ${data.targetName}`;

  if (data.type === 'BLOCK_IP') {
    const profile = db.attackerProfiles?.find((p) => p.ip === data.targetId || p.id === data.targetId);
    if (profile) {
      profile.threatLevel = 'Critical';
    }
  } else if (data.type === 'RESTRICT_MCP_TOOL') {
    const tool = db.mcpDecoys?.find((t) => t.id === data.targetId || t.name === data.targetId);
    if (tool) {
      tool.enabled = false;
    }
  } else if (data.type === 'ISOLATE_DECOY') {
    const hp = db.honeypots?.find((h) => h.id === data.targetId);
    if (hp) {
      hp.status = 'stopped';
    }
  }

  const auditBlock = appendAuditBlock(`CONTAINMENT_${data.type}`, {
    targetId: data.targetId,
    targetName: data.targetName,
    executedBy,
    details,
  });

  const newAction: ContainmentAction = {
    id: `act-${Date.now().toString(36)}`,
    type: data.type,
    targetId: data.targetId,
    targetName: data.targetName,
    status: 'EXECUTED',
    executedBy,
    timestamp: new Date().toISOString(),
    auditBlockHash: auditBlock.blockHash,
    details,
  };

  db.containmentActions.unshift(newAction);
  writeDb(db);

  return newAction;
}

File: backend/api/services/response_service.py
"""Response / containment service — port of src/server/services/response.service.ts"""
import time
from ..db import read_db, write_db, append_audit_block

def _now():
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()

def get_all_containment_actions() -> list:
    return read_db().get("containmentActions") or []

def execute_containment(data: dict) -> dict:
    db = read_db()
    if not db.get("containmentActions"): db["containmentActions"] = []
    executed_by = data.get("executedBy", "SOC Operator (CipherNest Desktop)")
    details = data.get("reason") or f"Automated containment rule executed for target: {data['targetName']}"
    action_type = data["type"]
    target_id = data["targetId"]

    if action_type == "BLOCK_IP":
        for p in db.get("attackerProfiles", []):
            if p["ip"] == target_id or p["id"] == target_id:
                p["threatLevel"] = "Critical"
    elif action_type == "RESTRICT_MCP_TOOL":
        for t in (db.get("mcpDecoys") or []):
            if t["id"] == target_id or t["name"] == target_id:
                t["enabled"] = False
    elif action_type == "ISOLATE_DECOY":
        for h in db.get("honeypots", []):
            if h["id"] == target_id:
                h["status"] = "stopped"

    audit = append_audit_block(f"CONTAINMENT_{action_type}", {
        "targetId": target_id,
        "targetName": data["targetName"],
        "executedBy": executed_by,
        "details": details,
    })
    action = {
        "id": f"act-{int(time.time()*1000):x}",
        "type": action_type,
        "targetId": target_id,
        "targetName": data["targetName"],
        "status": "EXECUTED",
        "executedBy": executed_by,
        "timestamp": _now(),
        "auditBlockHash": audit["blockHash"],
        "details": details,
    }
    db["containmentActions"].insert(0, action)
    write_db(db)
    return action
